### ISSUE TITLE
Jagged / padding version of torch.stack / torch.cat

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#65156]
This is a frequent primitive for collation: https://github.com/pytorch/pytorch/issues/1512

```python
import torch

def stack_jagged(tensors, fill_value = 0):
    # does not use F.pad to not allocate, could as well use F.pad's arg names instead 
    shape = [len(tensors)] + [max(t.shape[dim] for t in tensors) for dim in range(len(tensors[0].shape))]
    res = torch.full(shape, fill_value, dtype = tensors[0].dtype, device = tensors[0].device)
    for u, t in zip(res, tensors):
        u[tuple(map(slice, t.shape))] = t
    return u

if __name__ == '__main__':
    print(stack_jagged([torch.zeros(3, 4, 5), torch.zeros(3, 5, 6)]).shape)
```

Also, having a better indexing function than `u[tuple(map(slice, t.shape))] = t` would be nice - e.g. multi-dim narrow/slice function

**Comment by zou3519** [comment]
Is https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html what you're looking for?

**Comment by vadimkantorov** [comment]
Yes, but generalized beyond B x T x * format

**Comment by vadimkantorov** [comment]
One more real-world example: https://github.com/pytorch/vision/blob/ef711591a5db69d36f904ab5c39dec13627a58ad/torchvision/models/detection/transform.py#L208

**Comment by vadimkantorov** [comment]
Of course, a more complicated padding behaviors may be needed in some scenarios, but it should be good to at least have a basic version in the core

**Comment by vadimkantorov** [comment]
@cpuhrsch If nested tensor is being merged, some sort of padded stack may be a natural op to support

**Comment by cpuhrsch** [comment]
@vadimkantorov How about something like this:

```
tensors = [torch.randn(5, 12), torch.randn(5, 8), torch.randn(5, 3)]
nt = torch.nested_tensor(tensors)
t = nt.to_padded_tensor()
```
`t` then has shape `(3, 5, 12)` and `t[i, :, tensors[i].size(1)]` is equal to `nt[i]`.

Is this what you have in mind? Any ops you'd want the NestedTensor `nt` to support?

`size_divisible` is something the user would want to implement manually (for now) by expanding padding one of the input Tensors.

The number of padded elements can be determined via `nt.numel() / t.numel()`.

**Comment by vadimkantorov** [comment]
About padded tensor. In practice, size_divisible is very useful for faster CUDA algo selection/processing and more predictable allocator perf, and the number of padded (or not padded) dimensions is needed per batch element (and for images, this is even two-dimensional, since height and width are both padded) and not just in total. I can also imagine that sometimes more advanced padding modes need to be supported (e.g. provided by F.pad. also currently F.pad padding argument format is very unintuitive since it starts from the end and doesn't match the order of tensor dims). Another useful operation is unpad from a padded tensor back to nested tensor (taking the padding mask/ counts as input). Frequent cases: pad input images or ROI boxes.

Some other useful operations are related to operations/reductions with lists of equally-sized tensors: https://github.com/pytorch/pytorch/issues/27522 to avoid intermediate torch.stack allocation, especially for a few tensor inputs

This is also related to foreach-ops, instead of tensor lists, they could maybe be merged into existing ops with dispatch based on Tensor or NestedTensor passed in.

**Comment by cpuhrsch** [comment]
@vadimkantorov - I agree with you, I'm trying to find the right intersection between functionality that I know will land soon. 

There will also be functionality that converts from a padded tensor back to a NestedTensor.

We'll also have NestedTensor setup in core such that external contributions are easier to support.

Do you have a list of existing torch operators you'd like to see NestedTensor support for?

**Comment by vadimkantorov** [comment]
About ops - reductions over nested dim torch.sum/logsumexp in my issue (for 3-tensor inputs or 4-tensor inputs). For speed, it may be better to allocate-copy, but sometimes memory-wise it'd be better to avoid it. Ideally, it could be controlled by the user.

For foreach-ops, I wonder if there is any actual use of flattened parameters in sharding impls. If yes, then maybe a nestedtensor could be supported for checking if tensors are actually contiguous in memory to do a single kernel call - although maybe not a very needed optimization.

There is also a general nested lists of tensors input format for torch.stack that is supported by numpy

**Comment by cpuhrsch** [comment]
@vadimkantorov - ok to summarize you want

- torch.sum
- torch.logsumexp

**Comment by vadimkantorov** [comment]
@cpuhrsch Another op that could support nestedtensor/tensorlist is batched torch.unique (or corresponding indices) across a dim and similar ops that now could return a tensor list

Also it would be useful to have some free functions for conversions of different "padded" formats: flat tensor with split indices, padded tensor with recorded amount of padding, tensor lists, padded tensor with padding mask (hopefully supporting compression into a not introduced-yet bittensor)

**Comment by cpuhrsch** [comment]
@vadimkantorov - Thanks for raising this! Could I kindly ask you to open a separate issue for this so we can track this as well?

**Comment by vadimkantorov** [comment]
Not sure sparse is the right label...

**Comment by cpuhrsch** [comment]
@vadimkantorov - Yes and no. I also added the nestedtensor label.

**Comment by vadimkantorov** [comment]
About separate issue - I'm not sure what's the design idea of NestedTensor exactly and what's your design idea for conversion between different formats... E.g. do you wish to have several versions of TensorList depending on format?

for pad_stack a padded format makes most of sense, for some others TensorList may be best. but it may be good to support some variants. Maybe via some namedtuples that can be decomposed easily into plain tuples

**Comment by cpuhrsch** [comment]
@vadimkantorov by the way, NestedTensor is [now in core](https://pytorch.org/docs/master/nested.html) so you can do

```
>>> a, b = torch.arange(3), torch.arange(5) + 3
>>> a
tensor([0, 1, 2])
>>> b
tensor([3, 4, 5, 6, 7])
>>> nt = torch.nested_tensor([a, b])
>>> nt
nested_tensor([
  tensor([0, 1, 2]),
  tensor([3, 4, 5, 6, 7])
])
>>> nt.unbind()
(tensor([0, 1, 2]), tensor([3, 4, 5, 6, 7]))
```

**Comment by vadimkantorov** [comment]
i wonder could decompose into:
`t0, t1 = t.unbind()` be done without explicit call into unbind? similar to regular python's namedtuples?

**Comment by cpuhrsch** [comment]
@vadimkantorov we could support indexing `t1 = t[1]`

**Comment by vadimkantorov** [comment]
maybe namedtuples PaddedTensorWithPadInfo and PaddedTensorWithMask could be introduced that would contain the padded tensor and some padding info. If they are simple namedtuple, they would get decomposed easily as well. pad_stack (or F.pad applied to nestedtensro) could return one of these

**Comment by vadimkantorov** [comment]
Will it return back the padding information in some compact nice form? Or allow to compose a padding mask? It's often useful (since often padding / unpadding are needed in various parts of model)

Hmm, I see, so it's always copying. I guess for foreach methods nestedtensor supporting no-copy would be nice too (as it's already the case)

**Comment by cpuhrsch** [comment]
@vadimkantorov - we can return a NestedSize. It would look like this.
```
>>> nt = torch.nested_tensor([torch.randn(3, 351, 640), torch.randn(3, 480, 640), torch.randn(3, 640, 423)])
>>> nt.nested_size()
NestedSize([
	torch.Size([3, 351, 640]),
	torch.Size([3, 480, 640]),
	torch.Size([3, 640, 423]),
])
```

Next to `to_padded_tensor` we could also add a `to_tensor_mask` to generate a (padded, mask) tuple or `to_mask` to just return the mask (which we already have, just not publicly).


We can use the recycle the foreach kernels for various NestedTensor operations. Which operations do you need?

I agree that a `divisible_by` flag will be useful, but we have no core support to force a certain shape for e.g. optimal Tensor core utilization. One could however write a Tensor subclass that enforces this constraint via torch_dispatch.

**Comment by vadimkantorov** [comment]
It seems that divisible_by needs not any new special types. This kind of processsing already exists in detectron2 and all other vision frameworks. I also think that to_padded_tensor could be better served by a more generic `pad` method that can accept flexible padding options (pad left, pad right for each dim, type of padding)

**Comment by vadimkantorov** [comment]
> i wonder could decompose into: `t0, t1 = t.unbind()` be done without explicit call into unbind? similar to regular python's namedtuples?

btw this probably will be auto-supported if the nestedtensor supports enumeration (i.e. iteration over the first dim), just like regular tensor does already

**Comment by vadimkantorov** [comment]
@cpuhrsch maybe another usecase for TensorList (or even NestedTensor in this case, but I still like TensorLists as well :) ): https://github.com/pytorch/vision/pull/6278/files#diff-5288f1a12d4cbc77afac42d57b75796da3c8734c0fa4cda80e5c07ac5d16212dR273

**Comment by vadimkantorov** [comment]
@cpuhrsch Are jagged tensors equivalent of nestedtensors? https://twitter.com/PyTorch/status/1551977077984743424

**Comment by cpuhrsch** [comment]
@vadimkantorov - mostly, yes. cc @jbschlosser

**Comment by vadimkantorov** [comment]
Then why introduce a duplicate concept?.. although I think jagged is a better name than nested

**Comment by jbschlosser** [comment]
Hm it is a bit unfortunate that both terminologies are utilized. However, AFAIK the jagged terminology is only used in FBGEMM and NestedTensor is the concept within PyTorch Core.

> btw this probably will be auto-supported if the nestedtensor supports enumeration (i.e. iteration over the first dim), just like regular tensor does already

@vadimkantorov FYI enumeration is supported in nightly - NT implements [Python's `Mapping` protocol](https://docs.python.org/3/c-api/mapping.html) now for indexing, and we essentially got this for free :)

**Comment by vadimkantorov** [comment]
Yes, I’m mainly complaining about lack of synchronization on a very similar concept. Hopefully, fbgemm could use nested tensor as structures directly or influence its improvement.

About naming, I think, jagged array is more established terminology (in C# and java), so I find it a better naming over nested tensor.

**Comment by vadimkantorov** [comment]
@cpuhrsch if current memory format/copying will be codified somewhat, than matmul across the last dimension can be done without unbind/bind loop (i think unbind/bind is happening now)

**Comment by vadimkantorov** [comment]
Is fused cat+padding now implemented as `torch._chunk_cat`?

- https://github.com/pytorch/pytorch/pull/122888

**Comment by vadimkantorov** [comment]
Padded cat matches Example 1 in your issue, no? (modulo dim)
